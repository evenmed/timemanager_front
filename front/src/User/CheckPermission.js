import React, { useContext } from "react";
import PropTypes from "prop-types";
import { UserContext } from "./User";
import arrayIncludesAny from "../../lib/arrayIncludesAny";
import Error from "../helpers/Error";

/**
 * Checks if current user has a specific permission.
 * If not, can optionally show an error or simply not render children
 */
const CheckPermission = ({ permission, children, showError }) => {
  const { permissions: userPermissions } = useContext(UserContext);

  // Make sure user is signed in
  if (!userPermissions)
    return showError ? (
      <Error
        error="You must be signed in to
        access this page."
      />
    ) : null;

  // Check if user has necessary permission
  if (permission && !arrayIncludesAny(userPermissions, permission))
    return showError ? (
      <Error
        error="You are not authorized to
        access this page."
      />
    ) : null;

  return children;
};

CheckPermission.propTypes = {
  /**
   * Permission(s) to check for. If omitted, simply checks if user is logged in
   */
  permission: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  /**
   * If true, shows an error for unauthorized users.
   * If false it just doesn't render children.
   * Defaults to false.
   */
  showError: PropTypes.bool,
  children: PropTypes.node.isRequired,
};

export default CheckPermission;
