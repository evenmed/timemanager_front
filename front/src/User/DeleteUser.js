import React from "react";
import PropTypes from "prop-types";
import { useMutation, gql } from "@apollo/client";
import Error from "../helpers/Error";

const DELETE_USER_MUTATION = gql`
  mutation deleteUser($_id: ID!) {
    deleteUser(_id: $_id)
  }
`;

const DeleteUser = ({ userId, onSubmit: onCompleted }) => {
  const [deleteUser, { loading, error }] = useMutation(DELETE_USER_MUTATION, {
    variables: { _id: userId },
    refetchQueries: ["users"],
    onCompleted,
  });

  return (
    <>
      <Error error={error} />
      <button
        onClick={() =>
          confirm("Are you sure you want to delete this account?")
            ? deleteUser().catch((e) => console.error(e))
            : false
        }
        className="btn btn-danger btn-sm"
        type="button"
        disabled={loading}
      >
        <i className="fa fa-trash-alt"></i>{" "}
        {loading ? "Deleting..." : "Delete account"}
      </button>
    </>
  );
};

DeleteUser.propTypes = {
  /** ID of user that should be deleted */
  userId: PropTypes.string.isRequired,
  /** Callback for when a user is deleted properly */
  onSubmit: PropTypes.func,
};

export default DeleteUser;
