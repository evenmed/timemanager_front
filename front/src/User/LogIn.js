import { Formik } from "formik";
import { useMutation, gql } from "@apollo/client";

import Error from "../helpers/Error";

const LOG_IN_MUTATION = gql`
  mutation logIn($username: String!, $password: String!) {
    logIn(username: $username, password: $password) {
      username
    }
  }
`;

function LogIn(props) {
  const [logIn, { error, loading }] = useMutation(LOG_IN_MUTATION, {
    refetchQueries: ["me", "events"],
    awaitRefetchQueries: true,
  });

  return (
    <div>
      <h3>Log In</h3>
      <Formik
        initialValues={{ username: "", password: "" }}
        onSubmit={(values, { setSubmitting }) => {
          logIn({ variables: values })
            .then((res) => {
              setSubmitting(false);
            })
            .catch((e) => {
              setSubmitting(false);
              console.error(e);
            });
        }}
      >
        {({
          values,
          errors,
          handleChange,
          handleBlur,
          handleSubmit,
          isSubmitting,
          /* and other goodies */
        }) => (
          <form onSubmit={handleSubmit}>
            <Error error={error} />
            <div className="form-group">
              <label htmlFor="logInUsername">Username</label>
              <input
                type="username"
                className="form-control"
                name="username"
                id="logInUsername"
                autoComplete="username"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.username}
              />
            </div>
            <div className="form-group">
              <label htmlFor="logInPassword">Password</label>
              <input
                type="password"
                className="form-control"
                name="password"
                id="logInPassword"
                autoComplete="current-password"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.password}
              />
            </div>
            <button
              className="btn btn-success btn-block"
              type="submit"
              disabled={isSubmitting || loading}
            >
              {isSubmitting || loading ? "Submitting..." : "Submit"}
            </button>
          </form>
        )}
      </Formik>
    </div>
  );
}

export default LogIn;
