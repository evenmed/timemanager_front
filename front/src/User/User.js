import { useQuery, gql } from "@apollo/client";
import PropTypes from "prop-types";
import { createContext } from "react";
import Error from "../helpers/Error";

const CURRENT_USER_QUERY = gql`
  query me {
    me {
      _id
      username
      permissions
      preferredWorkTime
    }
  }
`;

const UserContext = createContext(false);

/**
 * Gets currently logged in user and makes it available via context.
 * Child must be a function that receives a isLoggedIn boolean
 */
const User = (props) => {
  const { loading, error, data } = useQuery(CURRENT_USER_QUERY);

  if (loading) return <p>Authenticating...</p>;
  if (error) return <Error error={error} />;

  return (
    <UserContext.Provider value={data.me}>
      {props.children(data.me ? true : false)}
    </UserContext.Provider>
  );
};

User.propTypes = {
  /** Function that receives a isLoggedIn boolean */
  children: PropTypes.func.isRequired,
};

export default User;
export { CURRENT_USER_QUERY, UserContext };
