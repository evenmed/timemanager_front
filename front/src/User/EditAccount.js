import React, { useContext } from "react";
import { useMutation, useQuery, gql } from "@apollo/client";
import { Formik, Field } from "formik";
import InputMask from "react-input-mask";
import PropTypes from "prop-types";

import { UserContext } from "./User";
import CheckPermission from "./CheckPermission";
import Error from "../helpers/Error";
import Loading from "../helpers/Loading";
import readableTimeString from "../../lib/readableTimeString";
import timeStringToMinutes from "../../lib/timeStringToMinutes";
import minutesToTimeString from "../../lib/minutesToTimeString";

const UPDATE_ACCOUNT_MUTATION = gql`
  mutation updateAccount(
    $userId: ID!
    $username: String!
    $preferredWorkTime: Int!
    $permissions: [Permission!]
  ) {
    updateAccount(
      userId: $userId
      username: $username
      preferredWorkTime: $preferredWorkTime
      permissions: $permissions
    ) {
      _id
    }
  }
`;

const SINGLE_USER_QUERY = gql`
  query user($_id: ID!) {
    user(_id: $_id) {
      _id
      username
      permissions
      preferredWorkTime
    }
  }
`;

const EditAccount = (props) => {
  const currentUser = useContext(UserContext);

  const [
    updateAccount,
    { error, loading },
  ] = useMutation(UPDATE_ACCOUNT_MUTATION, { refetchQueries: ["me", "users"] });

  const { data, error: queryError, loading: queryLoading } = useQuery(
    SINGLE_USER_QUERY,
    {
      variables: { _id: props.userId || currentUser._id },
      fetchPolicy: "network-only",
    }
  );

  if (queryLoading) return <Loading text="Fetching user..." />;
  if (queryError) return <Error error={error} />;

  const { user } = data;

  return (
    <Formik
      initialValues={{
        username: user.username,
        preferredWorkTime: minutesToTimeString(user.preferredWorkTime),
        permissions: user.permissions,
      }}
      validate={(values) => {
        const errors = {};

        if (values.preferredWorkTime) {
          const mins = timeStringToMinutes(values.preferredWorkTime);

          if (mins < 15 || mins > 1440)
            errors.preferredWorkTime =
              "Daily work objective must be between 15 mins and 24 hours";
        }

        return errors;
      }}
      validateOnChange={false}
      onSubmit={(values) =>
        updateAccount({
          variables: {
            username: values.username,
            preferredWorkTime: timeStringToMinutes(values.preferredWorkTime),
            userId: props.userId || user._id,
            // Only admins can edit permissions
            permissions: currentUser.permissions.includes("ADMIN")
              ? values.permissions
              : null,
          },
        })
          .then((res) => {
            if (props.onSubmit && res && res.data && res.data.updateAccount) {
              props.onSubmit();
            }
          })
          .catch((e) => {
            console.error(e);
          })
      }
    >
      {({
        values,
        errors,
        handleChange,
        handleBlur,
        handleSubmit,
        isSubmitting,
        /* and other goodies */
      }) => (
        <form onSubmit={handleSubmit}>
          <Error error={error} />
          <div className="form-group">
            <label htmlFor={`${user._id}-editUsername`}>Username</label>
            <input
              type="text"
              className="form-control"
              name="username"
              id={`${user._id}-editUsername`}
              autoComplete="username"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.username}
            />
            {errors.username && (
              <p className="text-danger">{errors.username}</p>
            )}
          </div>

          <div className="form-group">
            <label htmlFor={`${user._id}-editPwt`}>
              Daily work objective (HH:MM)
            </label>
            <InputMask
              mask="99:99"
              alwaysShowMask={true}
              className="form-control"
              name="preferredWorkTime"
              id={`${user._id}-editPwt`}
              autoComplete="off"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.preferredWorkTime}
              required
            />
            {readableTimeString(values.preferredWorkTime) && (
              <p className="text-muted mb-0">
                {readableTimeString(values.preferredWorkTime)}
              </p>
            )}
            {errors.preferredWorkTime && (
              <p className="text-danger">{errors.preferredWorkTime}</p>
            )}
          </div>

          <CheckPermission permission={["ADMIN"]}>
            <div>
              <strong>Permissions:</strong>
            </div>
            <div>
              <label>
                <Field type="checkbox" name="permissions" value="USER" /> USER
              </label>
            </div>
            <div>
              <label>
                <Field type="checkbox" name="permissions" value="USERMANAGER" />{" "}
                USERMANAGER
              </label>
            </div>
            <div>
              <label>
                <Field type="checkbox" name="permissions" value="ADMIN" /> ADMIN
              </label>
            </div>
          </CheckPermission>

          <button
            className="btn btn-success btn-block"
            type="submit"
            disabled={isSubmitting || loading}
          >
            {isSubmitting || loading ? "Saving..." : "Save"}
          </button>
        </form>
      )}
    </Formik>
  );
};

EditAccount.propTypes = {
  /** Callback for when an event is saved / deleted properly */
  onSubmit: PropTypes.func,
  /** ID of user to be edited (defaults to logged in user) */
  userId: PropTypes.string,
};

export default EditAccount;
export { SINGLE_USER_QUERY };
