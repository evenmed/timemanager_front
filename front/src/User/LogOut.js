import { useLazyQuery } from "@apollo/client";
import { CURRENT_USER_QUERY } from "./User";

/**
 * Log out button. Deletes session id from localStorage.
 */
const LogOut = (props) => {
  const [refreshUser, { error, loading }] = useLazyQuery(CURRENT_USER_QUERY, {
    fetchPolicy: "network-only",
  });

  if (error) console.error(error);

  return (
    <button
      disabled={loading}
      className="btn btn-danger"
      onClick={() => {
        if (props.onClick) props.onClick();

        localStorage.removeItem("token");
        refreshUser();
      }}
    >
      <i className="fa fa-sign-out"></i> Log out
    </button>
  );
};

export default LogOut;
