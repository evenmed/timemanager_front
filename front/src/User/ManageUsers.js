import React, { useState } from "react";
import { useQuery, gql } from "@apollo/client";
import Link from "next/link";
import minutesToHours from "../../lib/minutesToHours";
import Loading from "../helpers/Loading";
import Error from "../helpers/Error";
import EditAccountModal from "../Modals/EditAccountModal";
import CheckPermission from "./CheckPermission";

const USERS_QUERY = gql`
  query users($limit: Int!, $offset: Int!) {
    users(limit: $limit, offset: $offset) {
      _id
      username
      permissions
      preferredWorkTime
    }
  }
`;

const limit = 10; // Users to show per page

const ManageUsers = () => {
  const [offset, setOffset] = useState(0); // Users offset

  const { data, error, loading } = useQuery(USERS_QUERY, {
    variables: {
      limit: limit + 1, // We do this to check if we should show "next" btn
      offset,
    },
  });

  if (loading) return <Loading />;
  if (error) return <Error error={error} />;

  if (!data || !data.users || !data.users.length)
    return <Error error="No users found" />;

  const hasNextPage = data.users.length === limit + 1;

  const users = [...data.users];

  if (hasNextPage) users.pop();

  return (
    <div className="row pt-5">
      <div className="col-12">
        <EditAccountModal>
          {(showModal) => (
            <table className="table table-striped table-responsive-sm">
              <thead>
                <tr>
                  <th scope="col">Username</th>
                  <th scope="col">Daily work objective</th>
                  <th scope="col">Permissions</th>
                  <th scope="col">Actions</th>
                </tr>
              </thead>
              <tbody>
                {users.map((u) => (
                  <tr key={u._id}>
                    <td>{u.username}</td>
                    <td>{minutesToHours(u.preferredWorkTime)} hours</td>
                    <td>{u.permissions.join(", ")}</td>
                    <td className="p-1 text-center">
                      <button
                        onClick={() => showModal(u._id)}
                        className="btn btn-primary btn-sm m-1 text-nowrap"
                      >
                        <i className="fa fa-pencil" /> Edit
                      </button>
                      <CheckPermission permission="ADMIN">
                        <Link href={`calendar/${u._id}`}>
                          <a
                            className="btn btn-success m-1 btn-sm text-nowrap"
                            title={`${u.username}'s calendar`}
                          >
                            <i className="fa fa-calendar-alt" /> Calendar
                          </a>
                        </Link>
                      </CheckPermission>
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          )}
        </EditAccountModal>
      </div>
      <div className="col-12 text-center">
        <div className="btn-group" role="group" aria-label="Basic example">
          {offset > 0 && (
            <button
              onClick={() => setOffset(offset - limit)}
              type="button"
              className="btn btn-secondary"
            >
              <i className="fa fa-chevron-left"></i>
            </button>
          )}
          <span className="btn btn-secondary">
            Page {Math.round(offset / limit) + 1}
          </span>
          {hasNextPage && (
            <button
              onClick={() => setOffset(offset + limit)}
              type="button"
              className="btn btn-secondary"
            >
              <i className="fa fa-chevron-right"></i>
            </button>
          )}
        </div>
      </div>
    </div>
  );
};

export default ManageUsers;
