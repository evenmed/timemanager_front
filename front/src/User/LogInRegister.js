import { useState } from "react";
import LogIn from "./LogIn";
import Register from "./Register";

const LogInRegister = (props) => {
  // If true, show register form. Otherwise log in form.
  const [register, setRegister] = useState(false);

  return (
    <div className="mt-4">
      {register ? (
        <>
          <div className="mb-3 text-center">
            Already have an account?
            <br />
            <button
              className="btn btn-info mt-2"
              onClick={() => setRegister(!register)}
            >
              Log in
            </button>
          </div>
          <Register />
        </>
      ) : (
        <>
          <div className="mb-3 text-center">
            Don't have an account?
            <br />
            <button
              className="btn btn-info mt-2"
              onClick={() => setRegister(!register)}
            >
              Register
            </button>
          </div>
          <LogIn />
        </>
      )}
    </div>
  );
};

export default LogInRegister;
