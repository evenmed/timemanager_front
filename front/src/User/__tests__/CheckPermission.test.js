import CheckPermission from "../CheckPermission";
import { MockedProvider } from "@apollo/client/testing";
import { shallow, mount } from "enzyme";
import { act } from "react-dom/test-utils";
import toJSON from "enzyme-to-json";
import wait from "waait";
import User, { UserContext } from "../User";
import { fakeUser } from "../../../testUtils";

describe("<CheckPermission/>", () => {
  it("renders content for authorized users", () => {
    const wrapper = mount(
      <UserContext.Provider value={{ permissions: ["USER"] }}>
        <CheckPermission>
          <p className="content">Nice content</p>
        </CheckPermission>
      </UserContext.Provider>
    );

    const child = wrapper.find("p.content");
    expect(child).toHaveLength(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("doesn't render content for signed out users", () => {
    const wrapper = mount(
      <UserContext.Provider value={{}}>
        <CheckPermission>
          <p className="content">Nice content</p>
        </CheckPermission>
      </UserContext.Provider>
    );

    const child = wrapper.find("p.content");
    expect(child).toHaveLength(0);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders error for signed out users when showError is true", () => {
    const wrapper = mount(
      <UserContext.Provider value={{}}>
        <CheckPermission showError={true}>
          <p className="content">Nice content</p>
        </CheckPermission>
      </UserContext.Provider>
    );
    expect(wrapper.text()).toContain(
      "You must be signed in to access this page."
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it("renders error for unauthorized users when showError is true", () => {
    const wrapper = mount(
      <UserContext.Provider value={{ permissions: ["USER"] }}>
        <CheckPermission permission={"ADMIN"} showError={true}>
          <p className="content">Nice content</p>
        </CheckPermission>
      </UserContext.Provider>
    );
    expect(wrapper.text()).toContain(
      "You are not authorized to access this page."
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
