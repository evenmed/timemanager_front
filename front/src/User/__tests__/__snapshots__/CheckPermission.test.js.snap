// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckPermission/> doesn't render content for signed out users 1`] = `<CheckPermission />`;

exports[`<CheckPermission/> renders content for authorized users 1`] = `
<CheckPermission>
  <p
    className="content"
  >
    Nice content
  </p>
</CheckPermission>
`;

exports[`<CheckPermission/> renders error for signed out users when showError is true 1`] = `
<CheckPermission
  showError={true}
>
  <DisplayError
    error="You must be signed in to access this page."
  >
    <div
      className="alert alert-danger my-4"
      data-test="graphql-error"
    >
      <i
        className="fa fa-fw fa-exclamation-circle"
      />
       
      <strong>
        Error: 
      </strong>
      You must be signed in to access this page.
    </div>
  </DisplayError>
</CheckPermission>
`;

exports[`<CheckPermission/> renders error for unauthorized users when showError is true 1`] = `
<CheckPermission
  permission="ADMIN"
  showError={true}
>
  <DisplayError
    error="You are not authorized to access this page."
  >
    <div
      className="alert alert-danger my-4"
      data-test="graphql-error"
    >
      <i
        className="fa fa-fw fa-exclamation-circle"
      />
       
      <strong>
        Error: 
      </strong>
      You are not authorized to access this page.
    </div>
  </DisplayError>
</CheckPermission>
`;
